src/zenml/zen_stores/migrations/versions/7b651bf6822e_track_secrets_in_db.py

```diff
@@ -0,0 +1,94 @@
+"""track secrets in db [7b651bf6822e].
+
+Revision ID: 7b651bf6822e
+Revises: 1ac1b9c04da1
+Create Date: 2023-12-22 20:56:18.131906
+
+"""
+
+import sqlalchemy as sa
+from alembic import op
+
+from zenml.enums import StoreType
+from zenml.logger import get_logger
+
+logger = get_logger(__name__)
+
+# revision identifiers, used by Alembic.
+
+
+revision = "7b651bf6822e"
+down_revision = "1ac1b9c04da1"
+branch_labels = None
+depends_on = None
+
+
+def upgrade() -> None:
+    """Upgrade database schema and/or data, creating a new revision."""
+    from zenml.config.global_config import GlobalConfiguration
```

Comment:
I commented something similar on a cloudapi PR before where we decided importing these classes in migrations is okay as it will just run once.

I think here it is a little more complicated: This migration will also run in `zenml 0.75.x` in a few months/years from now, which means we need to keep those classes and all their (deprecated) methods around. Do you think this will lead to any trouble in the future?

Reply:
You're absolutely correct, we have to keep a lot of deprecated code around, even though it's only used in a DB migration script. Specifically, I'm talking about the deprecated "list_secrets" method in each secrets store implementation. We have two inter-related situations that I need to address here:

1. getting rid of deprecated code: this is a bit laborious, because I'd have to retest the migration again for all possible store types, but easy. I just have to move all those functions and their dependents inside the migration script.
2. getting rid of all imports in the migration script: technically impossible. These imports are strictly required to allow the migration script to connect to the external secrets store back-end. This needs to happen today, tomorrow, a few months from now or even a few years from now. As long as we have to support direct upgrades from 0.50.0 or earlier versions. On the bright side, we're only importing major concepts here: the global configuration and the SQL store. As long as we don't call attributes or methods prone to removal or deprecation on these objects (see the previous point), this should be safe.

Reply:
So the point is, we won't be able to remove those classes anyway as long as we want to support upgrading from pre-0.50 right? In that case I guess that's the best case we have, even though ideally this would not even be an automated migration in my opinion but rather something users do manually as there might be multiple reasons for this to fail.